docs/kb.md

```diff
@@ -0,0 +1,100 @@
+# Knowlege Base
+
+## Debugging
+
+By default, most of the code is running in graph mode (i.e. decoreated by
+tf.function directly or indirectly), which makes debugging difficult. When
+ debugging the unittests, you can include the following line to make the code
+ run in eager mode, which makes debugging much easier:
+ ```python
+ tf.config.experimental_run_functions_eagerly(True)
+ ```
+
+If you are debugging model running using alf.bin.train (i.e. with command `python -m alf.bin.train ...`),
+you can set the following gin configuration to run in eager mode:
+```
+TrainerConfig.use_tf_functions=False
+```
+
+### Debugging using VisualStudio Code
+
+Currently, ALF uses separate processes to launch multiple environments. Because
+vscode does not support debug for multiprocessing (https://github.com/microsoft/ptvsd/issues/1706),
+in order to debug in vscode, you need to make ALF not to start separate processes
+by setting the following gin config:
+```
+create_environment.num_parallel_environments=1
+PolicyTrainer._create_environment.nonparallel=True
+```
+
+## Algorithm
+
+Algorithm is the most important concept in ALF. TODO: more description about the design.
+
+## ActionTimeStep
+
+`ActionTimeStep` is a data structure extended from `TimeStep` of [TF-Agents](https://github.com/tensorflow/agents)
+for storing the information from the result of each environment step. It contains
+five fields:
+* `observation`: observation from the environment. It can be a nest of Tensors.
```

Comment:
It would be better to include when exactly is each field taken, e.g. observation is taken after the previous action and a world step; step_type is whether the current state (after previous action and environment step) is a first or middle or last state; discount is how much next step's value would contribute to this state.  (Not sure if this is accurate.)

This should clarify the sequence of things quite a bit.

Reply:
Added more detail